src/main/java/com/coveros/selenified/utilities/CombinedPDFReport.java

```diff
@@ -0,0 +1,194 @@
+                contents.close();
+
+                // Now add the link annotation, so the click on "Jump to page three" works
+                for (int reportIndex = 1; reportIndex <= testReportsList.size(); reportIndex++) {
+                    int pageIndex = reportIndex - 1;
+                    PDAnnotationLink pageLink = createLink(ph, font, fontSize,
+                            "Go to " + testReportsList.get(pageIndex), reportIndex);
+                    addLink(document, tableOfContents, pageLink, pageIndex);
+                }
+
+                try (PDDocument newDoc = new PDDocument()) {
+                    PDPageTree allPages = document.getDocumentCatalog().getPages();
+
+                    allPages.insertBefore(tableOfContents, allPages.get(0));
+                    for (PDPage page : allPages) {
+                        newDoc.addPage(page);
+                    }
+                    newDoc.save(outputDirectory + FILE_NAME);
+                }
+            } catch (Exception e) {
+                log.error(e);
+            }
+        }
+    }
+
+    /**
+     * Creates a map of the output directory of each suite paired with a list of all the ITestNGMethod objects
+     * in that suite
+     *
+     * @param suites the suites object from the TestNG tests executed
+     * @return Map<String, List<ITestNGMethod>>: a map of all output directories and their tests
+     */
+    protected Map<String, List<ITestNGMethod>> getTestSuiteMap(List<ISuite> suites) {
+        Map<String, List<ITestNGMethod>> testReportsMap = new HashMap<>();
+        for (ISuite iSuite : suites) {
+            String testReportDirectory = iSuite.getOutputDirectory();
+            List<ITestNGMethod> iTestNGMethods = iSuite.getAllMethods();
+            testReportsMap.put(testReportDirectory, iTestNGMethods);
+        }
+        return testReportsMap;
+    }
+
+    /**
+     * Creates a list of all test files, using the map to get the directory and test name
+     *
+     * @param testReportsMap the test report map object from getTestSuiteMap
+     * @return List<File>: a list of all test files in all suites
+     */
+    protected List<File> getAllTestReportFiles(Map<String, List<ITestNGMethod>> testReportsMap) {
+        List<File> testReportsList = new ArrayList<>();
+        for (Map.Entry<String, List<ITestNGMethod>> testReportsMapEntry : testReportsMap.entrySet()) {
+            String testReportDirectory = testReportsMapEntry.getKey();
+            for (ITestNGMethod iTestNGMethod : testReportsMapEntry.getValue()) {
+                File testReport = new File(testReportDirectory,
+                        iTestNGMethod.getTestClass().getName() + "." + iTestNGMethod.getMethodName() + ".pdf");
+                addToReport(testReportsList, testReport);
+            }
+        }
+        return testReportsList;
+    }
+
+    private void addToReport(List<File> testReportsList, File testReport) {
+        System.out.println(testReport.getAbsolutePath());
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Replace this use of System.out or System.err by a logger. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/sonar/coding_rules#rule_key=squid%3AS106)